design/resource-dependencies.md

```diff
@@ -0,0 +1,118 @@
+# Resource dependency relationships
+
+### Inventory of resources that could have dependency relationships
+All Rook-Ceph CRDs:
+- CephCluster
+- CephBlockPool
+- CephFilesystem
+- CephObjectStore
+- CephNFS
+- CephClient
+- CephRBDMirror
+- CephFilesystemMirror
+- CephObjectRealm
+- CephObjectZoneGroup
+- CephObjectZone
+
+Resources for which Rook-Ceph acts as a driver/provisioner:
+- ObjectBucketClaim (lib-bucket-provisioner)
+- ObjectBucket (COSI)
+
+Kubernetes resources which can depend on Rook:
+- StorageClass
+- PersistentVolume
+
+## Proposed dependency relationships
+
+#### `CephCluster`
+Primary dependents which can create/delete pools:
+- CephBlockPools in the same namespace
+- CephFilesystems in the same namespace
+- CephObjectStores in the same namespace
+
+Dependents which can consume arbitrary pools including user-created pools:
+- CephNFSes in the same namespace
+- CephClients in the same namespace
+
+#### `CephBlockPool`
+Dependents which can consume pools:
+- CephNFS in the same namespace that have `spec.pool == <a provider pool>`
+- CephClient in the same namespace that has any `spec.caps` value with the string `pool=<a provider pool>`
+
+Dependents via CSI:
+- StorageClasses that have:
+  - `provisioner == <operator-namespace>.rbd.csi.ceph.com` AND
+  - `parameters.clusterID == <ceph-cluster-namespace>` AND
+  - `parameters.pool` OR `parameters.dataPool` references a `CephBlockPool` pool
+- PersistentVolumes that have:
+  - `spec.CSI.Driver` == `<operator-namespace>.rbd.csi.ceph.com` AND
+  - `spec.CSI.VolumeAttributes["clusterID"] == <ceph-cluster-namespace>` AND
+  - `spec.CSI.VolumeAttributes["pool"]` OR `spec.CSI.VolumeAttributes["journalPool"]` references a `CephBlockPool` pool
+  - OR
+  - `spec.storageClassName == <name of StorageClass which references the CephBlockPool>`
+
+Dependents via FlexVolume:
+- StorageClasses that have:
+- PersistentVolumes that have:
+
+TODO: decide if we should treat PVs OR StorageClasses as dependents. IMO, in following with the rest
+of the doc, a StorageClass that exists and references a pool is a good dependent and has a more
+defined structure than PV's `VolumeAttributed` arbitrary k-v list which can change due to changes in
+ceph-csi. Users will not be able to delete the StorageClass until all PVs which reference it are
```

Comment:
I'm pretty sure you can delete a storage class even if there are PVs from it. I expect the storage class is only used at PV creation time.

Reply:
Oh! You're right. I guess we need to check both. :|